src/main/java/no/uio/ifi/localega/doa/aspects/AAIAspect.java

```diff
@@ -0,0 +1,106 @@
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.aspectj.lang.annotation.Around;
+import org.aspectj.lang.annotation.Aspect;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Component;
+
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Slf4j
+@Aspect
+@Component
+public class AAIAspect {
+
+    public static final String DATASETS = "datasets";
+
+    @Value("${ga4gh.passport.openid-configuration-url}")
+    private String openIDConfigurationURL;
+
+    @Value("${ga4gh.passport.public-key-path}")
+    private String passportPublicKeyPath;
+
+    @Value("${ga4gh.visa.public-key-path}")
+    private String visaPublicKeyPath;
+
+    @Autowired
+    protected HttpServletRequest request;
+
+    @Around("execution(public * no.uio.ifi.localega.doa.rest.*.*(..))")
+    public Object authenticate(ProceedingJoinPoint joinPoint) throws Throwable {
+        try {
+            String jwtToken = getJWTToken();
+            DecodedJWT decodedJWT = JWT.decode(jwtToken);
+            boolean isVisa = decodedJWT.getClaims().containsKey("ga4gh_visa_v1");
+            Collection<Visa> visas = new ArrayList<>();
+            if (isVisa) {
+                getVisa(jwtToken).ifPresent(visas::add);
+            } else {
+                visas.addAll(getVisas(jwtToken));
+            }
+            List<Visa> controlledAccessGrantsVisas = visas
+                    .stream()
+                    .filter(v -> v.getType().equalsIgnoreCase(VisaType.ControlledAccessGrants.name()))
+                    .collect(Collectors.toList());
+            log.info("Authentication and authorization attempt. User {} provided following valid GA4GH Visas: {}", decodedJWT.getSubject(), controlledAccessGrantsVisas);
+            Set<String> datasets = controlledAccessGrantsVisas
+                    .stream()
+                    .map(Visa::getValue)
+                    .map(d -> StringUtils.stripEnd(d, "/"))
+                    .map(d -> StringUtils.substringAfterLast(d, "/"))
+                    .collect(Collectors.toSet());
+            log.info("User has access to the following datasets: {}", datasets);
+            request.setAttribute(DATASETS, datasets);
+            return joinPoint.proceed();
+        } catch (Exception e) {
+            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(e.getMessage());
+        }
+    }
+
+    protected Collection<Visa> getVisas(String accessToken) {
+        Collection<String> visaTokens;
+        try {
+            String passportPublicKey = Files.readString(Path.of(passportPublicKeyPath));
```

Comment:
do I understand correctly that here we don't have the option to retrieve a JWK (or multiple) we can work with known public keys ?

Reply:
Yes, if the key can't be loaded (which most likely means that it was not provided), then we fall back to using JKU from the header of the Visa tokens.

Reply:
got it.